frontend/src/components/user/OpportunitySaveUserControl.tsx

```diff
@@ -0,0 +1,126 @@
+"use client";
+
+import { useUser } from "src/services/auth/useUser";
+
+import { useTranslations } from "next-intl";
+import Link from "next/link";
+import { useParams } from "next/navigation";
+import { useEffect, useRef, useState } from "react";
+import { ModalRef, ModalToggleButton } from "@trussworks/react-uswds";
+
+import { LoginModal } from "src/components/LoginModal";
+import SaveButton from "src/components/SaveButton";
+import { USWDSIcon } from "src/components/USWDSIcon";
+
+export const OpportunitySaveUserControl = () => {
+  const t = useTranslations("OpportunityListing");
+  const modalRef = useRef<ModalRef>(null);
+  const params = useParams();
+  const opportunity_id = String(params.id);
+
+  const { user } = useUser();
+  const [saved, setSaved] = useState(false);
+  const [showMessage, setshowMessage] = useState(false);
+  const [savedError, setSavedError] = useState(false);
+  const [loading, setLoading] = useState(false);
+
+  const closeMessage = () => {
+    setshowMessage(false);
+  };
+
+  const userSavedOppCallback = async () => {
+    setLoading(true);
+
+    const method = saved ? "DELETE" : "POST";
+    try {
+      const res = await fetch("/api/user/saved-opportunities", {
+        method,
+        headers: {
+          opportunity_id,
+        },
+      });
+      if (res.ok && res.status === 200) {
+        const data = (await res.json()) as { type: string };
+        data.type === "save" ? setSaved(true) : setSaved(false);
+      } else {
+        setSavedError(true);
+      }
+    } catch (error) {
+      setSavedError(true);
+      console.error(error);
+    } finally {
+      setshowMessage(true);
+      setLoading(false);
+    }
+  };
+
+  useEffect(() => {
+    if (!user?.token) return;
+    setLoading(true);
+    fetch(`/api/user/saved-opportunities/${opportunity_id}`)
+      .then((res) => (res.ok && res.status === 200 ? res.json() : []))
+      .then((data: { length: number }) => {
+        data.length ?? setSaved(true);
+      })
+      .finally(() => {
+        setLoading(false);
+      })
+      .catch((error) => {
+        console.error(error);
```

Comment:
There is no error state for the user visiting the page and an issue with fetching the status of the saved opportunity.